File: LeetCode-solution-Java/spiral_Matrix.java

/**
 *Given an m x n matrix, return all elements of the matrix in spiral order.

 

Example 1:


Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:


Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 

Constraints:

m == matrix.length
n == matrix[i].length
1 <= m, n <= 10
-100 <= matrix[i][j] <= 100 
 */
import java.util.*;;

public class spiral_Matrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();

        int a = matrix.length; // vertical length
        int b = matrix[0].length; // horizonal length

        int i = 0;
        int j = 0;

        int x = 1;

        int itemNumber = a * b;
        while (itemNumber > 0) {

            for (; j < b && itemNumber > 0; j++) {
                list.add(matrix[i][j]);
                itemNumber--;
            }
            j--;
            for (i = i + 1; i < a && itemNumber > 0; i++) {
                list.add(matrix[i][j]);
                itemNumber--;
            }
            i--;
            for (j = b - 2; j >= x-1 && itemNumber > 0; j--) {
                list.add(matrix[i][j]);
                itemNumber--;

            }
            j++;
            for (i = a - 2; i >= x && itemNumber > 0; i--) {
                list.add(matrix[i][j]);
                itemNumber--;

            }
            i++;

            x++;
            j++;

            a--;
            b--;

        }

        return list;
    }
}

File: Leetcode-Solution-Kotlin/spiralMatrix.kt
fun spiralOrder(matrix: Array<IntArray>): List<Int> {
    var m1 = 0
    var n1 = 0
    var m2 = matrix.size - 1
    var n2 = matrix[0].size - 1
    val ans = ArrayList<Int>()
    var items = matrix.size * matrix[0].size

    while (items > 0) {
        for (j in n1..n2) {
            if (items <= 0) break
            items--
            ans.add(matrix[m1][j])
        }
        for (i in m1 + 1..m2) {
            if (items <= 0) break
            items--
            ans.add(matrix[i][n2])
        }
        for (j in n2 - 1 downTo n1) {
            if (items <= 0) break
            items--
            ans.add(matrix[m2][j])
        }
        for (i in m2 - 1 downTo m1 + 1) {
            if (items <= 0) break
            items--
            ans.add(matrix[i][n1])
        }
        m2--
        n2--
        m1++
        n1++
    }
    return ans
}